torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -813,38 +813,52 @@ at::IntArrayRef strides_or_error(const Tensor & input, c10::string_view const &
     return input.strides();
   } else {
     return IntArrayRef({});
   }
 }
 
-Tensor mm_mat1_backward(const Tensor & grad, const Tensor & mat2, at::IntArrayRef mat1_sizes, at::IntArrayRef mat1_strides, const Scalar & alpha) {
-  // if input was column-major, return grad as column-order for efficiency
-  if (mat1_strides[0] == 1 && mat1_strides[1] == mat1_sizes[0]) {
-    return maybe_multiply(mat2.conj().mm(grad.t()).t(), alpha.conj());
-  } else {
-    return maybe_multiply(grad.mm(mat2.t().conj()), alpha.conj());
-  }
-}
-
-Tensor mm_mat2_backward(const Tensor & grad, const Tensor & mat1, IntArrayRef sizes, IntArrayRef strides, const Scalar & alpha) {
-  // if input was column-major, return grad as column-order for efficiency
-  if (strides[0] == 1 && strides[1] == sizes[0]) {
-    if (mat1.is_sparse()) {
-      // Since mm(dense, sparse) doesn't exist,
-      // pass a transposed output matrix to the underlying "addmm"
-      // function directly.
-      int64_t out_rows = mat1.size(1);
-      int64_t out_cols = grad.size(1);
-      Tensor t = at::zeros({}, grad.options()).expand({out_rows, out_cols}, true);
-      Tensor r = at::empty({out_cols, out_rows}, grad.options()).t();
-      at::addmm_out(r, t, mat1.t(), grad, alpha, 1);
-      return r;
+Tensor mm_mat1_backward(const Tensor& grad, const Tensor& mat2, at::IntArrayRef mat1_sizes, at::IntArrayRef mat1_strides, c10::Layout mat1_layout, const Scalar& alpha) {
+  if (grad.layout() == c10::kStrided && mat2.layout() == c10::kStrided && mat1_layout == c10::kStrided) {
+    // if input was column-major, return grad as column-order for efficiency
+    if (mat1_strides[0] == 1 && mat1_strides[1] == mat1_sizes[0]) {
+      return maybe_multiply(mat2.conj().mm(grad.t()).t(), alpha.conj());
     }
-    return maybe_multiply(grad.t().mm(mat1.conj()).t(), alpha.conj());
-  } else {
-    return maybe_multiply(mat1.t().conj().mm(grad), alpha.conj());
+  }
+
+  auto result = maybe_multiply(grad.mm(mat2.t().conj()), alpha.conj());
+
+  switch (mat1_layout) {
+    case c10::kStrided:
+      return result.to_dense();
+    case c10::kSparseCsr:
+      return result.to_sparse_csr();
+    default:
+      TORCH_CHECK(false, "mm_mat1_backward is not implemented for grad layout ", grad.layout(), ", mat2 layout ", mat2.layout(), ", mat1 layout ", mat1_layout);
+  }
+}
+
+Tensor mm_mat2_backward(const Tensor& grad, const Tensor& mat1, IntArrayRef mat2_sizes, IntArrayRef mat2_strides, c10::Layout mat2_layout, const Scalar& alpha) {
+  if (grad.layout() == c10::kStrided && mat1.layout() == c10::kStrided && mat2_layout == c10::kStrided) {
+    // if input was column-major, return grad as column-order for efficiency
+    if (mat2_strides[0] == 1 && mat2_strides[1] == mat2_sizes[0]) {
+      return maybe_multiply(grad.t().mm(mat1.conj()).t(), alpha.conj());
+    }
+  }
+
+  auto result = maybe_multiply(mat1.t().conj().mm(grad), alpha.conj());
+
+  switch (mat2_layout) {
```

Comment:
Can't you do `result.to(mat2_layout)`?

Reply:
No, it's not implemented. Related discussion in https://github.com/pytorch/pytorch/issues/46165
```py
In [1]: import torch
In [2]: a = torch.randn(3, 3)
In [3]: a.to(a.layout)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-d5bb2a40248e> in <module>
----> 1 a.to(a.layout)

TypeError: to() received an invalid combination of arguments - got (torch.layout), but expected one of:
 * (torch.device device, torch.dtype dtype, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 * (torch.dtype dtype, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 * (Tensor tensor, bool non_blocking, bool copy, *, torch.memory_format memory_format)
```